### PR TITLE
Security boot

### PR DESCRIPTION
**Pull request by DevIgork**

### CODE REVIEW

File: src/main/java/com/example/bookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity httpSecurity) throws Exception {
+        return httpSecurity
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/auth/**", "/error","/swagger-ui/**")
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/config/SecurityConfig.java]:
```suggestion
                                .requestMatchers("/auth/**", "/error","/swagger-ui/**", "/v3/api-docs/**")
```

File: src/main/java/com/example/bookstore/controller/BookController.java [hunk 2]
```diff
@@ -24,35 +25,40 @@
...
     }
 
+    @PreAuthorize("hasRole('USER')")
     @Operation(summary = "Get book", description = "get book by id")
     @GetMapping("/{id}")
     public BookDto getBookById(@PathVariable Long id) {
         return bookService.getBookById(id);
     }
 
+    @PreAuthorize("hasRole('ADMIN')")
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasAnyAuthority('ROLE_USER', 'ROLE_ADMIN')")
```

File: src/main/java/com/example/bookstore/controller/BookController.java [hunk 3]
```diff
@@ -24,35 +25,40 @@
 @Tag(name = "Book management", description = "Endpoints for managing products")
 @RequiredArgsConstructor
 @RestController
-@RequestMapping(value = "/books")
+@RequestMapping(value = "/api/books")
 public class BookController {
     private final BookService bookService;
 
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasAnyAuthority('ROLE_USER', 'ROLE_ADMIN')")
```

File: src/main/java/com/example/bookstore/model/User.java [hunk 4]
```diff
@@ -32,4 +41,46 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/com/example/bookstore/model/User.java [hunk 5]
```diff
@@ -32,4 +41,46 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/com/example/bookstore/model/User.java [hunk 6]
```diff
@@ -32,4 +41,46 @@ public class User {
...
+    }
+
+    @Override
+    public String getPassword() {
+        return password;
+    }
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/com/example/bookstore/repository/RoleRepository.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package com.example.bookstore.repository;
+
+import com.example.bookstore.model.Role;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/repository/RoleRepository.java:9]:
```suggestion
@Repository
public interface RoleRepository extends JpaRepository<Role, Long> {
```

File: src/main/resources/db/changelog/changes/03-create-roles-table.yaml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: role
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+                    unique: true
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/03-create-roles-table.yaml]: add empty line to the end of file

File: src/main/resources/db/changelog/changes/04-create-users-roles.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+                    deleteCascade: true
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
+                    references: roles(id)
+                    foreignKeyName: fk_users_roles_role_id
+                    nullable: falsе
+                    primaryKey: true
+                    deleteCascade: true
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/04-create-users-roles.yaml]: add empty line to the end of file

File: src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -23,7 +30,10 @@ public UserResponseDto register(
         if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
             throw new RegistrationException("Can't register user email is already registered");
         }
-        User savedUser = userRepository.save(userMapper.toModel(requestDto));
+        User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
+        user.setRoles(Set.of(roleRepository.findById(USER_ROLE_ID).get()));
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java]:
```suggestion
        user.setRoles(Set.of(roleRepository.findByRole(ROLE_NAME).get()));
```

File: src/main/resources/db/changelog/changes/07-insert-users-roles-table.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/07-insert-users-roles-table.yaml:1]: please use `valueComputed`

File: src/main/resources/db/changelog/changes/06-insert-users-table.yaml [hunk 11]
```diff
@@ -0,0 +1,21 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-users-table
+      author: Igor Korolevich
+      changes:
+        - insert:
+            tableName: users
+            columns:
+              - column: {name: email, value: "igorkorolj@gmail.com" }
+              - column: {name: password ,value: "$2a$10$bLhBOwKvdc8Qj6ADdR3PEOu9OH79WyY2d7e8L7qbaZ27td7fAfVx2"}
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/06-insert-users-table.yaml]:
```suggestion
              - column: {name: password, value: "$2a$10$bLhBOwKvdc8Qj6ADdR3PEOu9OH79WyY2d7e8L7qbaZ27td7fAfVx2"}
```

File: src/main/resources/db/changelog/changes/08-insert-books-table.yaml [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+              - column: {name: cover_image, value: "https://balka-book.com/files/2023/07_06/12_56/u_files_store_5_7.jpg"}
+        - insert:
+            tableName: books
+            columns:
+              - column: { name: title, value: "English is not easy" }
+              - column: { name: author, value: "Luci Gutiérrez" }
+              - column: { name: isbn, value: "9789669820228" }
+              - column: { name: price, value: "461" }
+              - column: { name: description, value: "If you want to finally start speaking English, but studying boring rules in textbooks drives you crazy, then this book is for you!" }
+              - column: { name: cover_image, value: "https://book-ye.com.ua/upload/resize_cache/iblock/21e/520_860_1/1f6d32c7_e05f_11e9_8121_000c29ae1566_73737abd_593d_11ed_8175_0050568ef5e6.jpg" }
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/08-insert-books-table.yaml]: please add empty line to the end of file

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but some changes should be done according to my comment

File: src/main/java/com/example/bookstore/controller/BookController.java (hunk 2 above)
Comment (DevIgork) [reply src/main/java/com/example/bookstore/controller/BookController.java]: I doubt it's necessary to use it in create, but I get it

**Review by DevIgork — COMMENTED** [review]
I forgot to add one commit name, i dont have jwt

**Review by yevheniisynytsia — APPROVED** [review]

**Pull request merged**